protocol-jackson/src/main/java/io/camunda/zeebe/protocol/jackson/record/AbstractRecord.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.protocol.jackson.record;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeInfo.As;
+import com.fasterxml.jackson.annotation.JsonTypeInfo.Id;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonTypeIdResolver;
+import io.camunda.zeebe.protocol.jackson.record.RecordBuilder.ImmutableRecord;
+import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.RecordType;
+import io.camunda.zeebe.protocol.record.RecordValue;
+import io.camunda.zeebe.protocol.record.RejectionType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@ZeebeStyle
+@JsonDeserialize(as = ImmutableRecord.class)
+public abstract class AbstractRecord<T extends RecordValue>
+    implements Record<T>, LazyJsonSerializable {
+  @Value.Default
+  @JsonTypeInfo(use = Id.CUSTOM, include = As.EXTERNAL_PROPERTY, property = "valueType")
+  @JsonTypeIdResolver(IntentTypeIdResolver.class)
+  @Override
+  public Intent getIntent() {
+    return Intent.UNKNOWN;
+  }
+
+  @Value.Default
+  @Override
+  public RecordType getRecordType() {
+    return RecordType.NULL_VAL;
+  }
+
+  @Value.Default
+  @Override
+  public RejectionType getRejectionType() {
+    return RejectionType.NULL_VAL;
+  }
+
+  @JsonTypeInfo(use = Id.CUSTOM, include = As.EXTERNAL_PROPERTY, property = "valueType")
+  @JsonTypeIdResolver(ValueTypeIdResolver.class)
+  @Override
+  public abstract T getValue();
+
+  /** @return itself as the object is immutable and can be used as is */
```

Comment:
:question: This was something I debated. Technically, as the object is immutable, there's no need to actually clone it. However, I think most people expect a new object (as in, by reference) when cloning, so this may break the contract. It's easy to replace with `new RecordBuilder<>().from(this).build()`, so let me know what you think.

Reply:
👍 I'd leave it as it is

Reading the [Object.clone() docs](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#clone()), it's breaking some of the requirements, but it also states that these requirements are not absolute and depend on the context of the specific class.

The only problem I see with this approach is that it allows modification of the shared properties. For example, this.value refers to another object that may have its properties altered by calling methods on this.clone().value. Given the context we are in, it's safe to assume that T is also always an immutable object and cannot have its properties altered. I therefore think it's fine to break the `x.clone() != x` requirement.

The advantage is clearly in performance, and although I'm not one for premature optimization, I think we should just use the immutable context as an advantage here and just return `this`.